R/simulated_annealing.R

```diff
@@ -0,0 +1,76 @@
+#' Feature selector that performs simmulated annealing to select features.
+#' 
+#' @description Algorithm randomly chooses a set of features, trains on them, scores the model.
+#'Then the algorithm slightly modifies the chosen features randomly and tests to see
+#'if the model improves. If there is improvement, the newer model is kept, if not the 
+#'algorithm tests to see if the worse model is still kept based on a acceptance 
+#'probability that decreases as iterations continue and if the model performs worse.
+#' 
+#' @param scorer  A custom user-supplied function that accepts X and y (as defined below) as input and returns the error of the datasets.
+#' @param X data frame of training features
+#' @param y data frame of training targets
+#' @param c control rate of feature perturbation
+#' @param iterations number of iterations
+#' @param bools If true function returns array of boolean values instead of column indicies
+#' 
+#' @return array of selected features
+#' @export
+#' 
+#' @examples
+#' 
+simulated_annealing <- function(scorer, X, y, c = 1, iterations = 100, bools = FALSE) {
+    
+    # Is `scorer` a function?
+    if (class(scorer) != "function") {
+    stop("Expected a function for `scorer`." )
+    }
+
+    # Do we have a data.frame or something compatible like tibble?
+    if (!any(class(X) == "data.frame")) {
+        stop("Expected a `data.frame` object for `data`.")
+    }
+
+    # Set mutate percentage
+    mutate <- 0.05
+    
+    # Obtain initial array of randomly selected features
+    ftr_all <- c(1:length(X))
+    ftr_old <- c()
+    while (length(ftr_old) == 0){
+        ftr_old <- (rbinom(length(X), 1, 0.5) != 0)
+    }
+    score_old <- scorer(cbind(X[ftr_old], y))
+    
+    # Iterate through new versions of selected features
+    for (i in 1:iterations){
+        ftr_new <- ftr_old
+        ftr_mutate <- sample(c(1:length(X)), size = ceiling(length(ftr_all)*mutate), replace = FALSE)
```

Comment:
Jacky Ho made me a comment that also could be useful for your code. I share it with you just in case you want to integrate it to your code (or take was a piece of advice), but I believe that it is not necessary right now because the short time-window for the project: *"Avoid abbreviations in variable names when they are not obvious. In this case, for instance, I don't know what no_ftr means."*

Reply:
Thanks, we can update these for Milestone 3